### PR TITLE
New post functionality

### PR DESCRIPTION
**Pull request by mateuszfelczak**
This feature branch provides a new functionality which is adding a new post.

### CODE REVIEW

File: core/.DS_Store
Comment (jacoor) [inline core/.DS_Store:1]: why is this file here? WHy are other of those files here? Please remove.

**Comment by jacoor** [comment]
Please revert all non-functional changes.

File: pyproject.toml [hunk 1]
```diff
@@ -20,6 +20,7 @@ django-timesince = "^2021.5.27"
 pytest = "^8.1.1"
 ruff = "^0.4.5"
 pytest-django = "^4.8.0"
+autopep8 = "^2.3.0"
```
Comment (jacoor) [inline pyproject.toml]: Why adding this? We use ruff.

File: core/models.py [hunk 2]
```diff
@@ -80,6 +80,7 @@ class Post(GenericModel):
         help_text="Hash of the title + content to prevent overwriting already saved post",
     )
     display_counter = models.IntegerField(default=0)
+    author = models.ForeignKey(User, on_delete=models.CASCADE)
```
Comment (swietlikm) [inline core/models.py]: why overwrite author?

File: core/urls.py [hunk 3]
```diff
@@ -1,10 +1,11 @@
 from django.urls import path
 
-from .views import PostDetailView, PostListView, PostSaveView, PostVoteView
+from .views import AddPostView, PostDetailView, PostListView, PostSaveView, PostVoteView
 
 urlpatterns = [
         path("post-list/", PostListView.as_view(), name="post-list"),
         path("post-detail/<int:pk>/", PostDetailView.as_view(), name="post-detail"),
+        path("add-post/", AddPostView.as_view(), name="add-post"),
```
Comment (swietlikm) [inline core/urls.py]: Maybe it's better to stay with CRUD naming convention: CreatePostView, post-create

File: core/views.py [hunk 4]
```diff
@@ -28,6 +30,31 @@ def get_object(self: "Post", queryset: QuerySet[Post] | None = None) -> Post:
...
+        kwargs["user"] = self.request.user
+        return kwargs
+
+    def form_valid(self: "AddPostView", form: PostForm) -> HttpResponse:
+        form.instance.author = self.request.user
+        return super().form_valid(form)
+
+    def get_context_data(self: "AddPostView", **kwargs: any) -> dict[str, any]:
+        context = super().get_context_data(**kwargs)
+        context["communities"] = Community.objects.filter(is_active=True)
```
Comment (swietlikm) [inline core/views.py:84]: this filtering is already implemented by default (with ActiveOnlyManager)

File: templates/post-add.html [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            <label for="{{ form.content.id_for_label }}" class="form-label">{{ form.content.label }}</label>
+            {{ form.content }}
+            {% for error in form.content.errors %}
+            <div class="invalid-feedback d-block">
+                {{ error }}
+            </div>
+            {% endfor %}
+        </div>
+
+        <button type="submit" class="btn btn-primary">Dodaj post</button>
```
Comment (swietlikm) [inline templates/post-add.html]: Please translate button name into english

Comment (swietlikm) [reply templates/post-add.html]: also, you could use form|crispy that would dramatically lower number of form lines, up to you.

File: core/urls.py (hunk 3 above)
Comment (jacoor) [reply core/urls.py]: Yes, please.

**Comment by jacoor** [comment]
@mateuszfelczak 
1. Please fix the tests. It is no use for us to do full code review when tests fail. 
2. Please clean up the files.

File: core/tests/test_adding_post.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    assert response.context["post"].title == data["title"]
+    assert response.context["post"].content == data["content"]
+
+
+def test_add_post_invalid(client: Client, user: User, community: Community) -> None:
+    data = {"title": ""}  # Missing content
+    client.force_login(user)
+    response = client.post(reverse("add-post"), data=data)
+    assert response.status_code == 200
+    assert b"This field is required." in response.content
```
Comment (jacoor) [inline core/tests/test_adding_post.py]: This could be slightly improved. It shows field is required, but we don't know which field. 
One way to do it would be, probably, to do something like. 
response.context.form.fields["title'].errors.0 == ...

File: core/tests/test_adding_post.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    client.force_login(user)
+    response = client.post(reverse("add-post"), data=data, follow=True)
+    assert response.status_code == 200
+    assert response.context["post"].author == user
+    assert response.context["post"].title == data["title"]
+    assert response.context["post"].content == data["content"]
+
+
+def test_add_post_invalid(client: Client, user: User, community: Community) -> None:
+    data = {"title": ""}  # Missing content
```
Comment (jacoor) [inline core/tests/test_adding_post.py]: this comment here is not necessary, the code is self explanatory.

File: core/tests/test_adding_post.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+
+def test_add_post_unauthorized(client: Client, community: Community) -> None:
+    data = {
+        "community": community.pk,
+        "title": "Test Post Title",
+        "content": "This is a test post content.",
+    }
+    response = client.post(reverse("add-post"), data=data)
+    # Assert redirection to login page (replace 302 with actual redirect code if different)
```
Comment (jacoor) [inline core/tests/test_adding_post.py]: Is this comment needed?

File: core/views.py [hunk 9]
```diff
@@ -63,6 +63,32 @@ def post(self: "PostDetailView", request: HttpRequest, pk: int) -> HttpResponse:
...
 
+class AddPostView(LoginRequiredMixin, CreateView):
+    model = Post
+    form_class = PostForm
+    template_name = "post-add.html"
+    login_url = "login"
+
+    def get_form_kwargs(self: "AddPostView") -> dict[str, any]:
+        kwargs = super().get_form_kwargs()
+        # Set initial community to None
```
Comment (jacoor) [inline core/views.py]: Is this comment needed? Code is self explanatory.

File: entrypoint.sh
Comment (jacoor) [inline entrypoint.sh:1]: Please revert this change and all other ACL changes.

File: templates/post-add.html [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+{% block content %}
+<div class="container mt-5">
+    <h1 class="mb-4">Dodaj nowy post</h1>
+
+    {% if form.errors %}
+    <div class="alert alert-danger" role="alert">
+        <ul class="mb-0">
+            {% for field in form %}
+            {% for error in field.errors %}
+            <li>{{ error }}</li>
```
Comment (jacoor) [inline templates/post-add.html]: Bit of indentation please, will be more readable.

Comment (jacoor) [reply templates/post-add.html]: Actually, this seems like a duplicate, no? Later on you display errors next to field.

File: templates/post-add.html [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+        </ul>
+    </div>
+    {% endif %}
+
+    <form method="post">
+        {% csrf_token %}
+
+        <div class="mb-3">
+            <label for="{{ form.community.id_for_label }}" class="form-label">{{ form.community.label }}</label>
+            {{ form.community }}
```
Comment (jacoor) [inline templates/post-add.html]: I kinda prefer using django forms (form.as...) but as you wish.

**Review by jacoor — CHANGES_REQUESTED** [review]

File: core/.DS_Store
Comment (tgrycuk) [reply core/.DS_Store:1]: done

File: pyproject.toml (hunk 1 above)
Comment (tgrycuk) [reply pyproject.toml]: done

File: core/models.py (hunk 2 above)
Comment (tgrycuk) [reply core/models.py]: fixed

File: core/urls.py (hunk 3 above)
Comment (tgrycuk) [reply core/urls.py]: done

File: core/tests/test_adding_post.py (hunk 6 above)
Comment (tgrycuk) [reply core/tests/test_adding_post.py]: done

File: core/tests/test_adding_post.py (hunk 7 above)
Comment (tgrycuk) [reply core/tests/test_adding_post.py]: clean

File: core/tests/test_adding_post.py (hunk 8 above)
Comment (tgrycuk) [reply core/tests/test_adding_post.py]: clean

File: core/views.py (hunk 9 above)
Comment (tgrycuk) [reply core/views.py]: clean

File: core/models.py [hunk 12]
```diff
@@ -159,7 +157,8 @@ def is_saved(self: "Post", user: User) -> bool:
     def get_comments(self: "Post") -> models.QuerySet:
         return self.children.all()
 
-    def get_comment_form(self: "Post") -> CommentForm:
+    def get_comment_form(self: "Post") -> any:
+        from .forms import CommentForm
         return CommentForm(initial={"parent_id": self.pk})
```
Comment (swietlikm) [inline core/models.py:162]: What happened here? Why was changed this function?

File: templates/base.html [hunk 13]
```diff
@@ -37,6 +37,9 @@
                         <li class="nav-item">
                             <a class="nav-link" href="{% url 'post-list' %}">Post List</a>
                         </li>
+                        <li class="nav-item">
+                            <a class="nav-link" href="{% url 'post-create' %}">Post Create</a>
```
Comment (swietlikm) [inline templates/base.html:41]: This link text should be "more user friendly" like Create new post or Create Post 
image

File: core/forms.py [hunk 14]
```diff
@@ -1,11 +1,35 @@
...
+        widgets: ClassVar[dict[str, forms.Widget]] = {
+            "community": forms.Select(attrs={"class": "form-select"}),
+            "title": forms.TextInput(attrs={"class": "form-control"}),
+            "content": forms.Textarea(attrs={"class": "form-control"}),
+        }
+
+    def __init__(self: "Post", *args: tuple, **kwargs: dict) -> None:
+        self.user = kwargs.pop("user", None)
+        super().__init__(*args, **kwargs)
+        self.fields["community"].queryset = Community.objects.filter(is_active=True)
```
Comment (swietlikm) [inline core/forms.py:33]: this filtering is already implemented by default (with ActiveOnlyManager)

File: core/forms.py [hunk 15]
```diff
@@ -1,11 +1,35 @@
...
+        model = Post
+        fields: ClassVar[list[str]] = ["community", "title", "content"]
+        widgets: ClassVar[dict[str, forms.Widget]] = {
+            "community": forms.Select(attrs={"class": "form-select"}),
+            "title": forms.TextInput(attrs={"class": "form-control"}),
+            "content": forms.Textarea(attrs={"class": "form-control"}),
+        }
+
+    def __init__(self: "Post", *args: tuple, **kwargs: dict) -> None:
+        self.user = kwargs.pop("user", None)
```
Comment (swietlikm) [inline core/forms.py:31]: there is no user field in the Post model, we are using author, also user field can be deleted

File: core/views.py [hunk 16]
```diff
@@ -63,6 +63,31 @@ def post(self: "PostDetailView", request: HttpRequest, pk: int) -> HttpResponse:
...
+class PostCreateView(LoginRequiredMixin, CreateView):
+    model = Post
+    form_class = PostForm
+    template_name = "core/post-create.html"
+    login_url = "login"
+
+    def get_form_kwargs(self: "PostCreateView") -> dict[str, any]:
+        kwargs = super().get_form_kwargs()
+        kwargs["initial"] = {"community": None}
+        kwargs["user"] = self.request.user
```
Comment (swietlikm) [inline core/views.py:75]: you are passing user to form but at the end you are not using it becase you put author on the form_valid()

File: templates/post-add.html (hunk 10 above)
Comment (tgrycuk) [reply templates/post-add.html]: done

File: core/forms.py [hunk 17]
```diff
@@ -1,11 +1,35 @@
...
+        attrs={
+            "rows": 3,
+            "class": "form-control bg-dark text-light",
+        },
+    ))
+    parent_id = forms.IntegerField(required=True, widget=forms.HiddenInput)
+
+
+class PostForm(forms.ModelForm):
+    user: forms.models.BaseModelForm | None
```
Comment (jacoor) [inline core/forms.py:19]: Nice one!

**Comment by jacoor** [comment]
Branch name is not following git-flow. Should be feature/new-post

File: core/tests/test_adding_post.py [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+
+
+def generate_random_password(length: int = 12) -> str:
+    characters = string.ascii_letters + string.digits + string.punctuation
+    return "".join(secrets.choice(characters) for _ in range(length))
+
+
+@pytest.fixture()
+def user(client: Client) -> User:
+    password = generate_random_password()
```
Comment (jacoor) [inline core/tests/test_adding_post.py:23]: Could be simple string. Just remark, not change request.

File: core/views.py [hunk 19]
```diff
@@ -63,6 +63,31 @@ def post(self: "PostDetailView", request: HttpRequest, pk: int) -> HttpResponse:
         return HttpResponse(html_content)
 
 
+class PostCreateView(LoginRequiredMixin, CreateView):
+    model = Post
+    form_class = PostForm
+    template_name = "core/post-create.html"
+    login_url = "login"
+
+    def get_form_kwargs(self: "PostCreateView") -> dict[str, any]:
```
Comment (jacoor) [inline core/views.py:72]: Again, nice one!

**Review by jacoor — APPROVED** [review]

**Review by swietlikm — CHANGES_REQUESTED** [review]
